core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -2369,39 +2369,12 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assert(scheduler.getShuffleDependencies(rddE) === Set(shuffleDepA, shuffleDepC))
   }
 
-  test("SPARK-17644: After one stage is aborted for too many failed attempts, subsequent stages" +
+  test("SPARK-17644: After one stage is aborted for too many failed attempts, subsequent stages " +
     "still behave correctly on fetch failures") {
-    // Runs a job that always encounters a fetch failure, so should eventually be aborted
```

Comment:
@skonto in answer to your question, here's an example of a method that runs a closure that seems to capture the enclosing test class and fails. I moved the definition of these methods out of the test method, but didn't help. Moving to the companion object did. Not sure what is going on underneath there, or whether you might have expected the closure cleaner to handle this case. I am not worried about it, but just pointing out a slightly more complex example.

There are more failures in the mllib module, coming soon ...

Reply:
Ok I will have a look. Do you have the output of the failure? Scala test does not report much. Btw in these tests what I noticed is that only the last one failed (failAfter "A job with one fetch failure should eventually succeed"), so not sure if it is the closure or anything else (need to debug it).

Reply:
Just that the task (the rdd.map call's argument) isn't serializable for the same reason that the LegacyAccumulatorWrapper failed -- captures the test class, which has an unserializable AssertionsHelper field in a scalatest superclass. The problem here is capturing the enclosing test class to begin with as it's not relevant.

Reply:
If you move 
```
object FailThisAttempt {
  val _fail = new AtomicBoolean(true)
}
```
outside tests as a top object, tests pass, no need to move the functions to the companion object. 

Btw the closure cleaner does not look into the body of the lambda to check if references of other objects create an issue. This is done only for the old closures. According to document we only checked for the return statements. Also Lambdas dont have outers by definition.

Regarding the LegacyAccumulatorWrapper there is no closure

Reply:
Yeah you've said it more correctly, it's really the body of the lambda capturing references, and it just happens to be to an enclosing class of the test method in this case.

This is the kind of the I wonder if we *do* have to do, like in the old closure cleaner. Because LMF closures capture far less to begin with, it's much less of an issue. I also remember that closure cleaning such things got dicey because it's not clear when it's just OK to null some object's field. It may not be possible to do reasonably.

Reply:
Yeah, I was also wondering if I had to implement this as well, but I feel people need to move to 2.12 with a different mindset as things have changed. Not sure if it is possible as well, so asked @LRytz in jira.

Reply:
I was able to reproduce and investigate this further but cleaning in 2.11 does not affect serialization in this case. I moved the check at the beginning of the clean method in the ClosureCleaner. Waiting for feedback from @lrytz on the bytecode stuff.

Reply:
@adriaanm thanks for that comment great to understand what is happening with the janino thing.
Here I am referring to object FailThisAttempt that has to be moved outside the function in the test case to make serialization work. So it seems in scala 2.11 serialization worked without even cleaning anything.
The similar local example I have is:

```
  test("external reference") {

    def runJobWithTemporaryFetchFailure: Unit = {

      object FailThisAttempt {
        val _fail = new AtomicBoolean(true)
      }

      val retC = new C1()
      ClosureCleaner.clean( () => { if(FailThisAttempt._fail.get()) println("dsdsds"); else println("dd"); 4 } )
    }
    runJobWithTemporaryFetchFailure
  }
```